### PR TITLE
feat: Implementing the fact that we need to first send a notification to add a friend

### PR DESCRIPTION
**Pull request by yassine04e**
Generalising the notification class by adding a sector, specific to profile or travels and adding a friend is done by an fsUid in the reposuitory level

### CODE REVIEW

File: app/src/main/java/com/github/se/travelpouch/ui/notifications/NotificationItem.kt [hunk 1]
```diff
@@ -175,36 +177,63 @@ fun handleInvitationResponse(
...
+                  listTravelViewModel.selectTravel(updatedContainer)
+                  Toast.makeText(context, "User added successfully!", Toast.LENGTH_SHORT).show()
+                },
+                { Toast.makeText(context, "Failed to add user", Toast.LENGTH_SHORT).show() })
+          }
+          Toast.makeText(context, responseMessage, Toast.LENGTH_SHORT).show()
+        },
+        onFailure = { Toast.makeText(context, "Failed to get travel", Toast.LENGTH_SHORT).show() })
+  } else {
+    profileViewModel.addFriend(
```
Comment (ArthurChalard) [inline app/src/main/java/com/github/se/travelpouch/ui/notifications/NotificationItem.kt:214]: I think that you always add the friend relation, even if the user declined the invitation to be friend. You should test if the response is positive or negative before this line.

File: app/src/main/java/com/github/se/travelpouch/ui/notifications/NotificationItem.kt [hunk 2]
```diff
@@ -175,36 +177,63 @@ fun handleInvitationResponse(
...
+                { updatedContainer ->
+                  listTravelViewModel.selectTravel(updatedContainer)
+                  Toast.makeText(context, "User added successfully!", Toast.LENGTH_SHORT).show()
+                },
+                { Toast.makeText(context, "Failed to add user", Toast.LENGTH_SHORT).show() })
+          }
+          Toast.makeText(context, responseMessage, Toast.LENGTH_SHORT).show()
+        },
+        onFailure = { Toast.makeText(context, "Failed to get travel", Toast.LENGTH_SHORT).show() })
+  } else {
```
Comment (ArthurChalard) [inline app/src/main/java/com/github/se/travelpouch/ui/notifications/NotificationItem.kt:213]: You should do an else if as we could add some other sectors in the future.

**Review by ArthurChalard — CHANGES_REQUESTED** [review]
The notification is sent when the user wants to add a friend and the friend relation is not directly created. Never the less, you don't really take the answer of the other person into account and always add the friend relation when he answer. Furthermore, the identity of the person who invite is hidden behind his fsUid. This is not very user friendly. I think that you wanted to use the email for the notification, but maybe it would even be better to show the username of the person.

Concerning the structure of the code, it is very hard to follow due to the quantity of layers and a bit of logic in the UI. I know that some were pre-existant, but your additions are not simplifying it.  Especially the addition of the sector in the notification. I think that it would have been easier to use some other notification type like INVITATION_FRIEND. As you already coded every thing like this, let's keep it, but please think of the maintainability of the code when writing. We will all gain time later in the project.
Also, you should add some hints for the review in the comments of your PR for the reviewer since you added a sector for example. Explain briefly why and the implication. This avoid the reviewer to understand the whole code before starting to review for real.

I tried to point out the errors I describe and added some Javadoc requests. I may have missed some places where Javadoc is required.

File: app/src/main/java/com/github/se/travelpouch/ui/profile/ModifyingProfileScreen.kt [hunk 3]
```diff
@@ -235,10 +248,24 @@ fun ModifyingProfileScreen(
...
+
+                                  notificationViewModel.sendNotification(
+                                      Notification(
+                                          notificationViewModel.getNewUid(),
+                                          senderUid = profile.value.fsUid,
+                                          receiverUid = friendUid,
+                                          travelUid = null,
+                                          content =
+                                              NotificationContent.FriendInvitationNotification(
+                                                  profile.value.fsUid),
```
[the remark this answers is not in the document]
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/ui/profile/ModifyingProfileScreen.kt:265]: Good catch, I did pay attention. I changed it

File: app/src/main/java/com/github/se/travelpouch/ui/notifications/NotificationItem.kt (hunk 1 above)
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/ui/notifications/NotificationItem.kt:214]: Yes, I changed it

File: app/src/main/java/com/github/se/travelpouch/ui/notifications/NotificationItem.kt (hunk 2 above)
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/ui/notifications/NotificationItem.kt:213]: Ok, this was updated. I went for a switch

File: app/src/main/java/com/github/se/travelpouch/model/notifications/NotificationContainer.kt [hunk 4]
```diff
@@ -70,3 +76,8 @@ enum class NotificationStatus {
   READ,
   UNREAD
 }
+
+enum class NotificationSector {
```
[the remark this answers is not in the document]
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/model/notifications/NotificationContainer.kt:80]: Done

File: app/src/main/java/com/github/se/travelpouch/model/notifications/NotificationContent.kt [hunk 5]
```diff
@@ -37,6 +37,15 @@ sealed class NotificationContent {
     }
   }
 
+  data class FriendInvitationNotification(
+      val userEmail: String,
+  ) : NotificationContent() {
+
+    override fun toDisplayString(): String {
```
[the remark this answers is not in the document]
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/model/notifications/NotificationContent.kt:44]: Done

File: app/src/main/java/com/github/se/travelpouch/model/notifications/NotificationContent.kt [hunk 6]
```diff
@@ -116,6 +125,17 @@ sealed class NotificationContent {
     }
   }
 
+  data class FriendInvitationResponseNotification(val email: String, val accepted: Boolean) :
```
[the remark this answers is not in the document]
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/model/notifications/NotificationContent.kt:128]: Done

File: app/src/main/java/com/github/se/travelpouch/model/profile/ProfileModelView.kt [hunk 7]
```diff
@@ -66,14 +66,14 @@ class ProfileModelView @Inject constructor(private val repository: ProfileReposi
   /**
    * This function allows us to add a friend to a user.
    *
-   * @param email (String) : The email of the friend to add
+   * @param fsUid (String) : The email of the friend to add
```
[the remark this answers is not in the document]
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/model/profile/ProfileModelView.kt:69]: Corrected

File: app/src/main/java/com/github/se/travelpouch/model/profile/ProfileRepository.kt [hunk 8]
```diff
@@ -50,12 +50,18 @@ interface ProfileRepository {
    *   failed
    */
   fun addFriend(
-      email: String,
+      fsUid: String,
```
[the remark this answers is not in the document]
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/model/profile/ProfileRepository.kt:53]: Done

**Comment by yassine04e** [comment]
To reply to your comment, I agree that tehcnically the username would be more user friendly, however, as we did not yet implement the unicity of a username, I thought emails are better

File: app/src/main/java/com/github/se/travelpouch/ui/profile/ModifyingProfileScreen.kt [hunk 9]
```diff
@@ -235,10 +248,24 @@ fun ModifyingProfileScreen(
...
+                                          notificationViewModel.getNewUid(),
+                                          senderUid = profile.value.fsUid,
+                                          receiverUid = friendUid,
+                                          travelUid = null,
+                                          content =
+                                              NotificationContent.FriendInvitationNotification(
+                                                  profile.value.email),
+                                          notificationType = NotificationType.INVITATION,
+                                          status = NotificationStatus.UNREAD,
+                                          sector = NotificationSector.PROFILE))
```
[the remark this answers is not in the document]
Comment (yassine04e) [reply app/src/main/java/com/github/se/travelpouch/ui/profile/ModifyingProfileScreen.kt:268]: Done.

**Comment by yassine04e** [comment]
Looking at the implementation. We have a problem with Notification. Indeed, there is no callback when sending a notification. This results in a toast being displayed without the guarantee that the notificarion is on the server. I don't want to increase the size of this PR, so if you agree, I will solve this issue for next sprint (and if possible, refactor in a different PR, how notifications are treated, to discard the sector mode).
